### PR TITLE
Enable secnetperf with xdp on azure VM

### PR DESCRIPTION
**Pull request by ami-GS**
## Description

Few fix to use secnetperf with xdp on two physically separated machines.

## Testing

Passed with command
`sudo pwsh ./scripts/secnetperf.ps1 -MsQuicCommit main -environment azure -plat linux -os ubuntu-22.04 -arch x64 -tls openssl3 -UserName root -io xdp`

## Documentation
N/A

### CODE REVIEW

File: scripts/secnetperf.ps1 [hunk 1]
```diff
@@ -225,19 +232,29 @@ if ($isWindows) {
...
+if ($useXDP -and $isWindows) { Install-XDP $Session $RemoteDir }
 if ($io -eq "wsk") { Install-Kernel $Session $RemoteDir $SecNetPerfDir }
 
 if (!$isWindows) {
     # Make sure the secnetperf binary is executable.
     Write-Host "Updating secnetperf permissions"
+    $GRO = "on"
+    $ENABLE_XDP=0
+    if ($io -eq "xdp") {
+        $GRO = "off"
```
Comment (nibanks) [inline scripts/secnetperf.ps1:244]: Why do you need to disable GRO to use XDP?

File: scripts/secnetperf.ps1 [hunk 2]
```diff
@@ -225,19 +232,29 @@ if ($isWindows) {
...
         $env:LD_LIBRARY_PATH = "${env:LD_LIBRARY_PATH}:$Using:RemoteDir/$Using:SecNetPerfDir"
         chmod +x "$Using:RemoteDir/$Using:SecNetPerfPath"
+        sudo sh -c "ethtool -K eth0 generic-receive-offload $Using:GRO"
+        $env:MSQUIC_ENABLE_XDP=$Using:ENABLE_XDP
     }
     $fullPath = Repo-Path $SecNetPerfDir
     $env:LD_LIBRARY_PATH = "${env:LD_LIBRARY_PATH}:$fullPath"
     chmod +x "./$SecNetPerfPath"
+    sudo sh -c "ethtool -K eth0 generic-receive-offload $GRO"
+    $env:MSQUIC_ENABLE_XDP=$ENABLE_XDP
```
Comment (nibanks) [inline scripts/secnetperf.ps1:257]: Reset this in cleanup?

File: src/platform/datapath_raw_xdp_linux.c [hunk 3]
```diff
@@ -357,7 +357,7 @@ AttachXdpProgram(struct xdp_program *Prog, XDP_INTERFACE *Interface, struct xsk_
         unsigned int xdp_flag;
     } AttachTypePairs[]  = {
         // { XDP_MODE_HW, XDP_FLAGS_HW_MODE },
-        { XDP_MODE_NATIVE, XDP_FLAGS_DRV_MODE },
+        // { XDP_MODE_NATIVE, XDP_FLAGS_DRV_MODE },
```
Comment (nibanks) [inline src/platform/datapath_raw_xdp_linux.c:360]: Why?

Comment (ami-GS) [reply src/platform/datapath_raw_xdp_linux.c:360]: DRV mode is not fully functional on Azure VM, but attachable to NIC without error. Normally there should be error if kenrel/hw doesn't support some features, then we can fallback to lower grade mode.
I'm asking accelerated networking team how to programmatically identify the functionality

File: scripts/secnetperf.ps1 (hunk 2 above)
Comment (ami-GS) [reply scripts/secnetperf.ps1:257]: cleanup phase (beginning of secnetperf.ps1) cannot know the default state of GRO.
So  I specify the feature off/on for xdp/others.

File: scripts/secnetperf.ps1 (hunk 1 above)
Comment (ami-GS) [reply scripts/secnetperf.ps1:244]: Decryption in connection.c start failing with XDP & GRO.
Maybe need additional implementation to split the merged chunks?

Comment (ami-GS) [reply scripts/secnetperf.ps1:244]: GRO might not work with generic XDP? (SKB_MODE)
image
https://conferences.sigcomm.org/sigcomm/2023/files/workshop-ebpf/1-CloudNative.pdf

and LRO (HW_GRO) does not work with XDP?
https://www.infradead.org/~mchehab/kernel_docs/networking/device_drivers/microsoft/netvsc.html

Comment (ami-GS) [reply scripts/secnetperf.ps1:244]: might be supported? need investigation. kernel version, library version, host/guest driver
https://lpc.events/event/11/contributions/939/attachments/771/1551/xdp-multi-buff.pdf

Comment (nibanks) [reply scripts/secnetperf.ps1:244]: I didn't see you disable GRO for all the other tests. Also, can't you programmatically disable it for your XDP socket instead of doing it system-wide?

File: scripts/secnetperf.ps1 (hunk 2 above)
Comment (nibanks) [reply scripts/secnetperf.ps1:257]: Is there no way to query the state at the beginning and save it? But really, as I said in the comment above, we should figure out a programmatic solution for just the XDP socket, instead of a system-wide setting.

File: src/platform/datapath_raw_xdp_linux.c (hunk 3 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp_linux.c:360]: Ok.

File: scripts/secnetperf.ps1 (hunk 1 above)
Comment (ami-GS) [reply scripts/secnetperf.ps1:244]: I realized that GRO in duonic is default off. And duonic support DRV mode attach.
DRV (driver) mode can reside with GRO because packet redirection by XDP program is "before" GRO.
DRV mode XDP doesn't GRO actually.
I tried 4 combination with duonic, SKB_MODE/DRV_MODE and GRO on/off. However the maximum ethernet frame size is always 1514. I don't actually see GROed packet. I saw larger packets on eth0 on Azure VM with GRO on then decryption failure.

As far as I checked ethtool code and document, it opens another socket by `socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);`. Then send payload, which is not well documented though, to kernel to set/unset flags. Document says that netlink feature is kind of ioctl so that netlink operation should be for each device (eth0 on Azure VM).
This socket (AF_NETLINK) is not associated with XDP socket. I believe it is difficult to disable/enable for each socket level. It is technically possible to do programmatically for system-wide, but it is kind of re-inventing the wheel of ethtool in msquic

File: scripts/secnetperf-helpers.psm1 [hunk 4]
```diff
@@ -228,6 +228,12 @@ function Cleanup-State {
             try { & "$Using:RemoteDir/scripts/log.ps1" -Cancel }
             catch { Write-Host "Failed to stop logging on server!" }
         }
+    } else {
+        # iterate all interface and "ip link set ${iface} xdp off"
+        $ifaces = ip link show | grep -oP '^\d+: \K[\w@]+' | cut -d'@' -f1
+        foreach ($iface in $ifaces) {
+            sudo ip link set $iface xdp off
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:235]: Can these commands fail? Should they be in a `try`/`catch` so one failure doesn't stop us from doing the rest?

Comment (ami-GS) [reply scripts/secnetperf-helpers.psm1:235]: Doesn't fail. "off" always succeeds

File: scripts/secnetperf.ps1 (hunk 2 above)
Comment (ami-GS) [reply scripts/secnetperf.ps1:257]: Possible to just query, but need external file to save the state for the case if the workload is killed during the run.

File: scripts/secnetperf-helpers.psm1 [hunk 5]
```diff
@@ -511,7 +529,7 @@ function Invoke-Secnetperf {
 
     # Start the server running.
     "> secnetperf $serverArgs" | Add-Content $serverOut
-    $job = Start-RemoteServer $Session "$RemoteDir/$SecNetPerfPath $serverArgs"
+    $job = Start-RemoteServer $Session "$RemoteDir/$SecNetPerfPath $serverArgs" (!$IsWindows -and $io -eq "xdp")
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:532]: 1. Since this condition never changes, let's make a `$sudo` variable above for this, equal to "" or "sudo MSQUIC_ENABLE_XDP=1".
2. Instead of changing the functions above, just inject `$sudo` here and below.
```suggestion
    $job = Start-RemoteServer $Session "$sudo $RemoteDir/$SecNetPerfPath $serverArgs"
```
Example change below in other comment.

File: scripts/secnetperf-helpers.psm1 [hunk 6]
```diff
@@ -521,7 +539,7 @@ function Invoke-Secnetperf {
         Write-Host "==============================`nRUN $($try+1):"
         "> secnetperf $clientArgs" | Add-Content $clientOut
         try {
-            $process = Start-LocalTest $clientPath $clientArgs $artifactDir
+            $process = Start-LocalTest $clientPath $clientArgs $artifactDir (!$IsWindows -and $io -eq "xdp")
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:542]:
```suggestion
            $process = Start-LocalTest "$sudo$clientPath" $clientArgs $artifactDir (!$IsWindows -and $io -eq "xdp")
```

File: scripts/secnetperf-helpers.psm1 [hunk 7]
```diff
@@ -291,6 +297,7 @@ function Start-LocalTest {
     } else {
         # We use bash to execute the test so we can collect core dumps.
         $pinfo.FileName = "bash"
+        $sudo = ""
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:300]: Unnecessary?

File: scripts/secnetperf-helpers.psm1 [hunk 8]
```diff
@@ -483,6 +490,11 @@ function Invoke-Secnetperf {
         continue
     }
 
+    $sudo=""
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:493]:
```suggestion
    $sudo = ""
```

File: scripts/secnetperf-helpers.psm1 [hunk 9]
```diff
@@ -483,6 +490,11 @@ function Invoke-Secnetperf {
...
+    if ((!$IsWindows -and $io -eq "xdp")) {
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:494]:
```suggestion
    if (!$IsWindows -and $io -eq "xdp") {
```

File: scripts/secnetperf-helpers.psm1 [hunk 10]
```diff
@@ -521,11 +533,14 @@ function Invoke-Secnetperf {
...
         "> secnetperf $clientArgs" | Add-Content $clientOut
         try {
-            $process = Start-LocalTest $clientPath $clientArgs $artifactDir
+            $process = Start-LocalTest "$sudo$clientPath" $clientArgs $artifactDir
             $rawOutput = Wait-LocalTest $process $artifactDir ($io -eq "wsk") 30000
             Write-Host $rawOutput
             $values[$tcp] += Get-TestOutput $rawOutput $metric
             if ($extraOutput) {
+                if ($sudo -ne "") {
+                    sudo chown $UserName $extraOutput
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:542]: Why is this necessary?

Comment (ami-GS) [reply scripts/secnetperf-helpers.psm1:542]: The extraOutput (latency.txt) is owned by root when running secnetperf with sudo. deleting the file from normal user get failed. Existence of latency.txt make next run fail

File: scripts/secnetperf-helpers.psm1 [hunk 11]
```diff
@@ -483,6 +489,9 @@ function Invoke-Secnetperf {
         continue
     }
 
+     # Linux XDP requires env var and sudo for now
+    $sudo = (!$IsWindows -and $io -eq "xdp") ? "sudo MSQUIC_ENABLE_XDP=1 " : ""
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:493]: Daiki, you had `sudo -E` here, but I replaced it with the above because I thought that was actually what you needed, right? Please confirm before merging.

Comment (nibanks) [reply scripts/secnetperf-helpers.psm1:493]: Oh, I see, you do it in the other script with `$env:`. I'll change back to `-E`

File: scripts/secnetperf-helpers.psm1 [hunk 12]
```diff
@@ -483,6 +489,9 @@ function Invoke-Secnetperf {
         continue
     }
 
+     # Linux XDP requires sudo for now
+    $sudo = (!$IsWindows -and $io -eq "xdp") ? "sudo -E " : ""
```
Comment (nibanks) [inline scripts/secnetperf-helpers.psm1:493]: Can you explain why the `-E` needs to be here? As I understand it, it gives edit permissions? Is that for the output artifacts?

File: scripts/secnetperf.ps1 [hunk 13]
```diff
@@ -235,19 +242,29 @@ if ($isWindows) {
 Configure-DumpCollection $Session
 
 # Install any dependent drivers.
-if ($useXDP) { Install-XDP $Session $RemoteDir }
+if ($useXDP -and $isWindows) { Install-XDP $Session $RemoteDir }
```
Comment (nibanks) [inline scripts/secnetperf.ps1:245]: This just made me realize, we won't be able to differentiate between 'Windows QUIC + XDP' and 'Linux QUIC + XDP' because we name them 'QUIC+XDP' in the data file (I think). @ProjectsByJackHe do we need to make the `$io` names unique?

**Review by nibanks — APPROVED** [review]

File: scripts/secnetperf.ps1 [hunk 14]
```diff
@@ -235,19 +242,29 @@ if ($isWindows) {
...
         $env:LD_LIBRARY_PATH = "${env:LD_LIBRARY_PATH}:$Using:RemoteDir/$Using:SecNetPerfDir"
+        $env:MSQUIC_ENABLE_XDP=$Using:ENABLE_XDP
         chmod +x "$Using:RemoteDir/$Using:SecNetPerfPath"
+        sudo sh -c "ethtool -K eth0 generic-receive-offload $Using:GRO"
     }
     $fullPath = Repo-Path $SecNetPerfDir
     $env:LD_LIBRARY_PATH = "${env:LD_LIBRARY_PATH}:$fullPath"
+    $env:MSQUIC_ENABLE_XDP=$ENABLE_XDP
     chmod +x "./$SecNetPerfPath"
+    sudo sh -c "ethtool -K eth0 generic-receive-offload $GRO"
```
Comment (nibanks) [inline scripts/secnetperf.ps1:267]: This seems to be failing: https://github.com/microsoft/netperf/actions/runs/8867963327/job/24347253651
```
Error: ethtool: bad command line argument(s)

OriginInfo            : netperf-peer
Exception             : System.Management.Automation.RemoteException: ethtool: 
                        bad command line argument(s)
TargetObject          : ethtool: bad command line argument(s)
CategoryInfo          : NotSpecified: (ethtool: bad command line argument(s):St
                        ring) [], RemoteException
FullyQualifiedErrorId : NativeCommandError
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 6
PipelineIterationInfo : {}

ethtool: bad command line argument(s)
```

File: scripts/secnetperf-helpers.psm1 (hunk 12 above)
Comment (ami-GS) [reply scripts/secnetperf-helpers.psm1:493]: sudo reset environment variables.
If my understanding is correct, LD_LIBRARY_PATH is not propagated to sudo execution.

**Comment by nibanks** [comment]
I see this on the 20.04 lab machine run of netperf ([link](https://github.com/microsoft/netperf/actions/runs/8873638329/job/24361315839)):
```
sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
```

**Comment by ami-GS** [comment]
netperf repo is monitoring each msquic PRs?

**Comment by nibanks** [comment]
> netperf repo is monitoring each msquic PRs?

Yep. Here is the latest run for your PR: https://github.com/microsoft/netperf/actions/runs/8882454145

File: src/platform/datapath_xplat.c [hunk 15]
```diff
@@ -49,6 +49,7 @@ CxPlatDataPathInitialize(
     char* EnableXdp = getenv("MSQUIC_ENABLE_XDP");
     if (EnableXdp != NULL && EnableXdp[0] == '1') {
 #endif
+    if (Config && Config->Flags & QUIC_EXECUTION_CONFIG_FLAG_XDP) {
```
Comment (nibanks) [inline src/platform/datapath_xplat.c:52]: Please update Windows XDP to use this flag too.

File: src/platform/datapath_xplat.c [hunk 16]
```diff
@@ -49,6 +49,7 @@ CxPlatDataPathInitialize(
     char* EnableXdp = getenv("MSQUIC_ENABLE_XDP");
```
Comment (nibanks) [inline src/platform/datapath_xplat.c:49]: Please remove env variable usage, and make sure Windows and Linux both use the flag below.

File: src/inc/msquic.h [hunk 17]
```diff
@@ -268,6 +268,7 @@ typedef enum QUIC_EXECUTION_CONFIG_FLAGS {
     QUIC_EXECUTION_CONFIG_FLAG_QTIP             = 0x0001,
     QUIC_EXECUTION_CONFIG_FLAG_RIO              = 0x0002,
 #endif
+    QUIC_EXECUTION_CONFIG_FLAG_XDP              = 0x0004,
```
[the remark this answers is not in the document]
Comment (ami-GS) [reply src/inc/msquic.h:271]: DataPathTest.cpp cannot see `QUIC_API_ENABLE_PREVIEW_FEATURES`

[the remark this answers is not in the document]
Comment (nibanks) [reply src/inc/msquic.h:271]: It needs to set `QUIC_API_ENABLE_PREVIEW_FEATURES` then. This needs to be hidden for now.

File: src/tools/spin/spinquic.cpp [hunk 18]
```diff
@@ -1538,6 +1538,20 @@ void start() {
...
+            QUIC_EXECUTION_CONFIG ExecutionConfig { QUIC_EXECUTION_CONFIG_FLAG_XDP, 10000, 0, {0} };
+            if (QUIC_FAILED(
+            Status =
+            MsQuic->SetParam(
+                nullptr,
+                QUIC_PARAM_GLOBAL_EXECUTION_CONFIG,
+                sizeof(ExecutionConfig),
+                &ExecutionConfig))) {
+                printf("Failed to set execution config %d\n", Status);
+            }
```
Comment (nibanks) [inline src/tools/spin/spinquic.cpp:1552]:
```suggestion
            if (QUIC_FAILED(
                Status =
                MsQuic->SetParam(
                    nullptr,
                    QUIC_PARAM_GLOBAL_EXECUTION_CONFIG,
                    sizeof(ExecutionConfig),
                    &ExecutionConfig))) {
                printf("Failed to set execution config %d\n", Status);
            }
```

File: src/platform/unittest/DataPathTest.cpp [hunk 19]
```diff
@@ -18,6 +18,7 @@
 #endif
 
 extern bool UseDuoNic;
+#define QUIC_EXECUTION_CONFIG_FLAG_XDP (QUIC_EXECUTION_CONFIG_FLAGS)0x0004
```
Comment (nibanks) [inline src/platform/unittest/DataPathTest.cpp:21]: Instead of this, simply define `QUIC_API_ENABLE_PREVIEW_FEATURES` above, before all the includes.

**Review by nibanks — DISMISSED** [review]

**Review by nibanks — APPROVED** [review]

File: src/platform/unittest/DataPathTest.cpp [hunk 20]
```diff
@@ -427,7 +427,9 @@ QuicAddr DataPathTest::UnspecIPv4;
 QuicAddr DataPathTest::UnspecIPv6;
 
 struct CxPlatDataPath {
-    QUIC_EXECUTION_CONFIG DefaultExecutionConfig { QUIC_EXECUTION_CONFIG_FLAG_NONE, 10000, 0, {0} };
+    static uint8_t RawConfig[QUIC_EXECUTION_CONFIG_MIN_SIZE + 256 * sizeof(uint16_t)] {0};
+    static QUIC_EXECUTION_CONFIG* DefaultExecutionConfig;
```
Comment (nibanks) [inline src/platform/unittest/DataPathTest.cpp:431]: I don't understand why you made this change. The old stuff should work. You definitely don't need to fully specify the list of processors.

Comment (ami-GS) [reply src/platform/unittest/DataPathTest.cpp:431]: `PollingIdleTimeoutUs == 10000` is for QTIP only?

Comment (ami-GS) [reply src/platform/unittest/DataPathTest.cpp]: OK this value is the root cause

**Review by nibanks — APPROVED** [review]
